azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -13,47 +13,180 @@
 # limitations under the License.
 #
 # Requires Python 2.6+ and Openssl 1.0+
-
-import time
+import threading
+from datetime import datetime as dt
 
 from azurelinuxagent.common import logger
-from azurelinuxagent.common.cgroup import CpuCgroup, MemoryCGroup
+
 
 class CGroupsTelemetry(object):
     """
     """
-    _tracked = {}
+    _tracked = []
+    _cgroup_metrics = {}
+    _rlock = threading.RLock()
 
     @staticmethod
-    def metrics_hierarchies():
-        return CGroupsTelemetry._hierarchies
+    def _get_metrics_list(metric):
+        return [metric.average(), metric.min(), metric.max(), metric.median(), metric.count(), metric.first_poll_time(),
+                metric.last_poll_time()]
+
+    @staticmethod
+    def _process_cgroup_metric(cgroup_metrics):
+        current_memory_usage, max_memory_levels, current_cpu_usage = cgroup_metrics.get_metrics()
+        processed_extension = {
+            "memory": {
+                "cur_mem": CGroupsTelemetry._get_metrics_list(current_memory_usage),
+                "max_mem": CGroupsTelemetry._get_metrics_list(max_memory_levels)
+            },
+            "cpu": {
+                "cur_cpu": CGroupsTelemetry._get_metrics_list(current_cpu_usage)
+            }
+        }
+        return processed_extension
 
     @staticmethod
     def track_cgroup(cgroup):
         """
-        Adds the given item to the list of tracked cgroups
+        Adds the given item to the dictionary of tracked cgroups
         """
-        pass  # TODO
+        with CGroupsTelemetry._rlock:
+            CGroupsTelemetry._tracked.append(cgroup)
+            logger.info("Started tracking new cgroup: {0}, path: {1}".format(cgroup.name, cgroup.path))
 
     @staticmethod
-    def is_tracked(name):
+    def is_tracked(name, controller):
         """
         Returns true if the given item is in the list of tracked items
+        O(n) operation. But limited to few cgroup objects we have.
         """
-        pass  # TODO
+        for cgroup in CGroupsTelemetry._tracked:
```

Comment:
You need to hold the lock while iterating thru _tracked

Reply:
This would cause locking issues in the `poll_all_tracked`. The `poll_all_tracked` has to lock to make sure it is polling data on the latest stable list. The `CGroupsTelemetry.stop_tracking(cgroup)` would hit the lock acquisition failure.

Current implementation:
```
def poll_all_tracked():
    with CGroupsTelemetry._rlock:
        for cgroup in CGroupsTelemetry._tracked[:]:
            if cgroup.name not in CGroupsTelemetry._cgroup_metrics:
                CGroupsTelemetry._cgroup_metrics[cgroup.name] = CgroupMetrics()
            CGroupsTelemetry._cgroup_metrics[cgroup.name].add_new_data(cgroup.controller, cgroup.collect())

            if not cgroup.is_active():
                CGroupsTelemetry.stop_tracking(cgroup)
                CGroupsTelemetry._cgroup_metrics[cgroup.name].marked_for_delete = True
```

If you want the locking to only happen in primitive operations such as `stop_tracking` and `track_cgroup`, It will still work.